salt/utils/boto3.py

```diff
@@ -0,0 +1,298 @@
+        cache_id('myinstance', 'i-a1b2c3')
+        instance_id = cache_id('myinstance')
+    '''
+    return partial(cache_id, service)
+
+
+def get_connection(service, module=None, region=None, key=None, keyid=None,
+                   profile=None):
+    '''
+    Return a boto connection for the service.
+
+    .. code-block:: python
+
+        conn = __utils__['boto.get_connection']('ec2', profile='custom_profile')
+    '''
+
+    module = module or service
+
+    cxkey, region, key, keyid = _get_profile(service, region, key,
+                                             keyid, profile)
+    cxkey = cxkey + ':conn'
+
+    if cxkey in __context__:
+        return __context__[cxkey]
+
+    try:
+        session = boto3.session.Session(aws_access_key_id=keyid,
+                          aws_secret_access_key=key,
+                          region_name=region)
+        if session is None:
+            raise SaltInvocationError('Region "{0}" is not '
+                                      'valid.'.format(region))
+        conn = session.client(module)
+        if conn is None:
+            raise SaltInvocationError('Region "{0}" is not '
+                                      'valid.'.format(region))
+    except boto.exception.NoAuthHandlerFound:
+        raise SaltInvocationError('No authentication credentials found when '
+                                  'attempting to make boto {0} connection to '
+                                  'region "{1}".'.format(service, region))
+    __context__[cxkey] = conn
+    return conn
+
+
+def get_connection_func(service, module=None):
+    '''
+    Returns a partial `get_connection` function for the provided service.
+
+    ... code-block:: python
+
+        get_conn = __utils__['boto.get_connection_func']('ec2')
+        conn = get_conn()
+    '''
+    return partial(get_connection, service, module=module)
+
+
+def get_error(e):
+    # The returns from boto modules vary greatly between modules. We need to
+    # assume that none of the data we're looking for exists.
+    aws = {}
+    if hasattr(e, 'status'):
+        aws['status'] = e.status
+    if hasattr(e, 'reason'):
+        aws['reason'] = e.reason
+    if hasattr(e, 'message') and e.message != '':
+        aws['message'] = e.message
+    if hasattr(e, 'error_code') and e.error_code is not None:
+        aws['code'] = e.error_code
+
+    if 'message' in aws and 'reason' in aws:
+        message = '{0}: {1}'.format(aws['reason'], aws['message'])
+    elif 'message' in aws:
+        message = aws['message']
+    elif 'reason' in aws:
+        message = aws['reason']
+    else:
+        message = ''
+    r = {'message': message}
+    if aws:
+        r['aws'] = aws
+    return r
+
+
+def exactly_n(l, n=1):
+    '''
+    Tests that exactly N items in an iterable are "truthy" (neither None,
+    False, nor 0).
+    '''
+    i = iter(l)
+    return all(any(i) for j in range(n)) and not any(i)
+
+
+def exactly_one(l):
```

Comment:
Most of the content of utils.boto3 is redundant with utils.boto. (It was really kind of a 'clone and own' approach. Some has been altered to fit boto3, some didn't need to be altered. I left the redundant code in place thinking that eventually boto2 would be fully replaced and utils/boto.py would be retired. Of course, the redundant stuff could be moved at that point. I'm open to changing it. Actually, I don't think I make use of these functions anywhere, anyway.